Share return value handling in trim analysis

sbomer opened the pull request:
This shares more of the HandleCall logic across ILLink/ILC/ILLink.RoslynAnalyzer. The return value handling has been moved from each tool's implementation (ReflectionMethodBodyScanner/TrimAnalysisVisitor) into the shared HandleCallAction, and the extra handling in MethodBodyScanner has been removed.

Addresses the suggestions from https://github.com/dotnet/runtime/pull/101212#issuecomment-2062529793 and https://github.com/dotnet/runtime/pull/101212#issuecomment-2063089448.

sbomer committed:
Share logic for handling return values

sbomer committed:
Remove MethodBodyScanner return value handling

jtschuster commented on src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:
```diff
@@ -27,7 +27,53 @@ internal partial struct HandleCallAction
+			if (!TryHandleIntrinsic (calledMethod, instanceValue, argumentValues, intrinsicId, out maybeMethodReturnValue))
+				HandleSharedIntrinsic (calledMethod, instanceValue, argumentValues, intrinsicId, out maybeMethodReturnValue);
+
+			// If we get here, we handled this as an intrinsic.  As a convenience, if the code above
+			// didn't set the return value (and the method has a return value), we will set it to be an
+			// unknown value with the return type of the method.
```
I know this is just moved from existing code, but ideally we'd want the intrinsic handling above to always set this rather than setting a default here, right? If so we should consider filing an issue to track it.

jtschuster reviewed (approved):
Thank you!

sbomer committed:
Merge remote-tracking branch 'origin/main' into intrinsicsCleanup

sbomer committed:
Avoid throwing from analyzer in Release builds

And add missing cases to the switch statement for the analyzer.

sbomer committed:
Fix ILC callsites

sbomer committed:
Fix Array_CreateInstance case

The PInvoke logic and the CheckAndReportRequires logic
should not be needed for this intrinsic, and the old shared
HandleCallAction would go to the default case that sets the
return value to Top and returns false.

Now this instead returns true and lets the shared return
value logic kick in.

MichalStrehovsky reviewed (approved):
LGTM otherwise, thanks!

sbomer committed:
PR feedback

- Return instead of out param in HandleCall
- FIx formatting
